app/src/main/java/com/amaze/filemanager/filesystem/FilenameHelper.kt

```diff
@@ -0,0 +1,239 @@
+
+        var start: Int = startArg
+
+        if (strip) {
+            filename = stripIncrementInternal(filename, removeRawNumbers)
+            dirname = stripIncrementInternal(dirname, removeRawNumbers)
+            basename = strip(basename, removeRawNumbers)
+        }
+
+        var retval = HybridFile(
+            file.mode,
+            dirname,
+            filename,
+            file.isDirectory(AppConfig.getInstance())
+        )
+
+        while (retval.exists(AppConfig.getInstance())) {
+            filename = if (extension.isNotBlank()) {
+                format(platform, basename, start++) + ".$extension"
+            } else {
+                format(platform, basename, start++)
+            }
+            retval = HybridFile(
+                file.mode,
+                dirname,
+                filename,
+                file.isDirectory(AppConfig.getInstance())
+            )
+        }
+
+        return retval
+    }
+
+    private fun stripIncrementInternal(input: String, removeRawNumbers: Boolean = false): String {
+        val source = StringBuilder().run {
+            append(REGEX_SOURCE)
+            if (removeRawNumbers) {
+                append(REGEX_RAW_NUMBERS)
+            }
+            toString()
+        }
+        return Regex("($source)+$", RegexOption.IGNORE_CASE).replace(input, "")
+    }
+
+    private fun stem(filepath: String, removeRawNumbers: Boolean = false): String {
+        val extension = filepath.pathFileExtension()
+        return stripIncrementInternal(
+            filepath.pathBasename().substringBefore(".$extension"),
+            removeRawNumbers
+        )
+    }
+
+    private fun ordinal(n: Int): String {
+        var retval = ordinals.getOrNull(((n % 100) - 20) % 10)
+        if (retval == null) {
+            retval = ordinals.getOrNull(n % 100)
+        }
+        if (retval == null) {
+            retval = ordinals[0]
+        }
+        return retval
+    }
+
+    // TODO: i18n
+    private fun format(flag: FilenameFormatFlag, stem: String, n: Int): String {
+        return when (flag) {
+            FilenameFormatFlag.DARWIN -> {
+                if (n == 1) {
+                    "$stem copy"
+                } else if (n > 1) {
+                    "$stem copy $n"
+                } else {
+                    stem
+                }
+            }
+            FilenameFormatFlag.LINUX -> {
+                when (n) {
+                    0 -> {
+                        stem
+                    }
+                    1 -> {
+                        "$stem (copy)"
+                    }
+                    2 -> {
+                        "$stem (another copy)"
+                    }
+                    else -> {
+                        "$stem (${toOrdinal(n)} copy)"
+                    }
+                }
+            }
+            // Windows and default formatting are the same.
+            else -> {
+                if (n > 1) {
```

Comment:
Here is the second problem. Let's say `startArg` initial value is 1. When `increment()` function initially call `format()` , it'll simply return the same filename because `n <= 1`. Then `increment()` will call `format()` for n = 2 and it'll work fine. I don't know if there's specific reason to add this `if (n>1)` condition but in my case this is creating the problem.

Reply:
I think it's understandable, if looking at the unit test for windows.

https://github.com/TeamAmaze/AmazeFileManager/blob/bda61488a4e75e3d37e21de85a094ed47c9ec35b/app/src/test/java/com/amaze/filemanager/filesystem/FilenameHelperWindowsIncrementNameTest.kt#L40-L48

Couldn't comment if it's right or wrong, as I was doing a direct port... but if I understand correctly, looks like you'd want to have filenames from starting with `file (2).txt` to `file (1).txt`, which is different from the original node.js implementation.

If that is the case, see if 1f2cd85 fits your needs.

Reply:
Yes, now it's working fine.